File: packages/python-sdk/src/opensend/events.py
"""Events resource for the OpenSend Python SDK."""

from __future__ import annotations

from typing import Optional, cast

from ._http import HttpClient
from ._types import (
    CreateEventPayload,
    CustomEvent,
    CustomEventListResponse,
    ListOptions,
    SendCustomEventResponse,
    SendEventPayload,
)


class EventsResource:
    """Create custom event schemas and send event deliveries (/api/events)."""

    def __init__(self, client: HttpClient) -> None:
        self._client = client

    def create(self, payload: CreateEventPayload) -> CustomEvent:
        """Define a new custom event schema."""
        return cast(CustomEvent, self._client.request("POST", "/api/events", payload))

    def list(self, options: Optional[ListOptions] = None) -> CustomEventListResponse:
        """List all custom event definitions."""
        opts = options or {}
        query: dict[str, str] = {}
        if opts.get("limit") is not None:
            query["limit"] = str(opts["limit"])
        if opts.get("after"):
            query["after"] = opts["after"]  # type: ignore[assignment]
        return cast(
            CustomEventListResponse,
            self._client.request("GET", "/api/events", params=query or None),
        )

    def send(self, payload: SendEventPayload) -> SendCustomEventResponse:
        """Send (fire) a custom event, resuming any matching automation runs."""
        return cast(
            SendCustomEventResponse,
            self._client.request("POST", "/api/events/send", payload),
        )

File: packages/go-sdk/events.go
package opensend

import (
	"context"
	"fmt"
	"net/http"
)

// EventsClient provides access to the /api/events endpoints.
type EventsClient struct {
	c *Client
}

// Create registers a new custom event schema via POST /api/events.
func (e *EventsClient) Create(ctx context.Context, req CreateEventRequest) (*CustomEvent, error) {
	var out CustomEvent
	if err := e.c.doJSON(ctx, http.MethodPost, "/api/events", req, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List retrieves registered event schemas via GET /api/events.
func (e *EventsClient) List(ctx context.Context, opts ListOptions) (*CustomEventListResponse, error) {
	params := map[string]string{}
	if opts.Limit != nil {
		params["limit"] = fmt.Sprintf("%d", *opts.Limit)
	}
	if opts.After != "" {
		params["after"] = opts.After
	}

	var out CustomEventListResponse
	if err := e.c.doJSON(ctx, http.MethodGet, buildQuery("/api/events", params), nil, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Send fires a custom event via POST /api/events/send.
func (e *EventsClient) Send(ctx context.Context, req SendEventRequest) (*SendCustomEventResponse, error) {
	var out SendCustomEventResponse
	if err := e.c.doJSON(ctx, http.MethodPost, "/api/events/send", req, RequestOptions{}, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
